File: tests/fixtures/tree-sitter/go/httpbin_client.go
package httpbin

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

const DefaultBaseURL = "https://httpbin.org"

type Doer interface {
	Do(req *http.Request) (*http.Response, error)
}

type Client struct {
	baseURL    string
	httpClient Doer
	timeout    time.Duration
}

type AnythingResponse struct {
	Args    map[string]string `json:"args"`
	Data    string            `json:"data"`
	Headers map[string]string `json:"headers"`
	Method  string            `json:"method"`
	URL     string            `json:"url"`
}

func NewClient(httpClient Doer, baseURL string) *Client {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	if baseURL == "" {
		baseURL = DefaultBaseURL
	}
	return &Client{
		baseURL:    baseURL,
		httpClient: httpClient,
		timeout:    10 * time.Second,
	}
}

func (c *Client) GetJSON(ctx context.Context, query map[string]string) (AnythingResponse, error) {
	req, err := c.newRequest(ctx, http.MethodGet, "/get", query, nil)
	if err != nil {
		return AnythingResponse{}, err
	}
	return c.decode(req)
}

func (c *Client) PostJSON(ctx context.Context, payload map[string]any) (AnythingResponse, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return AnythingResponse{}, err
	}
	req, err := c.newRequest(ctx, http.MethodPost, "/post", nil, bytes.NewReader(body))
	if err != nil {
		return AnythingResponse{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	return c.decode(req)
}

func (c Client) Status(ctx context.Context, code int) error {
	req, err := c.newRequest(ctx, http.MethodGet, fmt.Sprintf("/status/%d", code), nil, nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != code {
		return fmt.Errorf("httpbin status %d", resp.StatusCode)
	}
	return nil
}

func (*Client) Version() string {
	return "fixture-v1"
}

func (c *Client) decode(req *http.Request) (AnythingResponse, error) {
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return AnythingResponse{}, err
	}
	defer resp.Body.Close()
	var out AnythingResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return AnythingResponse{}, err
	}
	return out, nil
}

func (c *Client) newRequest(ctx context.Context, method string, endpoint string, query map[string]string, body io.Reader) (*http.Request, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	u, err := url.Parse(c.baseURL)
	if err != nil {
		return nil, err
	}
	u.Path = endpoint
	values := u.Query()
	for key, value := range query {
		values.Set(key, value)
	}
	u.RawQuery = values.Encode()
	req, err := http.NewRequestWithContext(ctx, method, u.String(), body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	return req, nil
}

File: tests/fixtures/tree-sitter/ts/httpbinClient.ts
export interface HttpTransport {
  fetch(input: Request | URL | string, init?: RequestInit): Promise<Response>
}

export type HttpBinPayload = {
  args: Record<string, string>
  data?: string
  headers: Record<string, string>
  method: string
  url: string
}

export type RequestOptions = {
  query?: Record<string, string>
  headers?: Record<string, string>
  timeoutMs?: number
  onRetry?: (attempt: { count: number; status?: number }) => void
}

export type Decoder = (response: Response, shape: { expectJson: boolean; endpoint: string }) => Promise<HttpBinPayload>

export class HttpBinClient {
  readonly baseUrl = "https://httpbin.org"
  private readonly token: string | undefined
  protected retryCount = 1

  constructor(private readonly transport: HttpTransport, token?: string) {
    this.token = token
  }

  decorateRequest = (request: Request, metadata: { endpoint: string; attempt: number }) => request

  async getJson(options: RequestOptions = {}): Promise<HttpBinPayload> {
    const request = this.buildRequest("/get", options)
    return this.send(request, { expectJson: true, endpoint: "/get" })
  }

  async postJson(payload: { message: string; tags: string[] }, options: RequestOptions = {}): Promise<HttpBinPayload> {
    const request = this.buildRequest("/post", {
      ...options,
      headers: { ...options.headers, "Content-Type": "application/json" }
    }, JSON.stringify(payload))
    return this.send(request, { expectJson: true, endpoint: "/post" })
  }

  public async status(code: number): Promise<boolean> {
    const request = this.buildRequest(`/status/${code}`, {})
    const response = await this.transport.fetch(request)
    return response.status === code
  }

  private buildRequest(endpoint: string, options: RequestOptions, body?: string): Request {
    const url = new URL(endpoint, this.baseUrl)
    for (const [key, value] of Object.entries(options.query ?? {})) {
      url.searchParams.set(key, value)
    }
    const init: RequestInit = {
      method: body === undefined ? "GET" : "POST",
      headers: this.createHeaders(options.headers)
    }
    if (body !== undefined) {
      init.body = body
    }
    const request = new Request(url, init)
    return this.decorateRequest(request, { endpoint, attempt: 1 })
  }

  protected createHeaders(headers: Record<string, string> = {}): Headers {
    const merged = new Headers(headers)
    if (this.token) {
      merged.set("Authorization", `Bearer ${this.token}`)
    }
    return merged
  }

  private async send(request: Request, decoderOptions: { expectJson: boolean; endpoint: string }): Promise<HttpBinPayload> {
    const response = await this.transport.fetch(request)
    return decodeHttpBin(response, decoderOptions)
  }
}

export async function decodeHttpBin(response: Response, shape: { expectJson: boolean; endpoint: string }): Promise<HttpBinPayload> {
  if (!shape.expectJson) {
    throw new Error(`unexpected non-json response for ${shape.endpoint}`)
  }
  return response.json() as Promise<HttpBinPayload>
}

export namespace HttpBinEndpoint {
  export function status(code: number): string {
    return `/status/${code}`
  }

  export const anything = (name: string) => `/anything/${name}`
}

const fetchTransport: HttpTransport = {
  fetch: (input, init) => fetch(input, init)
}

class RetryClock {
  now(): number {
    return Date.now()
  }
}

export { HttpBinClient as Client, RetryClock as Clock }

export default (transport: HttpTransport = fetchTransport) => new HttpBinClient(transport)
